rs/ledger_suite/icp/ledger/src/main.rs

```diff
@@ -745,11 +748,59 @@ fn main() {
     })
 }
 
+// We use 8MiB buffer
+const BUFFER_SIZE: usize = 8388608;
+
 fn post_upgrade(args: Option<LedgerCanisterPayload>) {
     let start = dfn_core::api::performance_counter(0);
-    let mut stable_reader = stable::StableReader::new();
+
+    // In order to read the first bytes we need to use ic_cdk.
+    // dfn_core assumes the first 4 bytes store stable memory length
+    // and return bytes starting from the 5th byte.
+    let mut magic_bytes_reader = ic_cdk::api::stable::StableReader::default();
+    const MAGIC_BYTES: &[u8; 3] = b"MGR";
+    let mut first_bytes = [0u8; 3];
+    let memory_manager_found = match magic_bytes_reader.read_exact(&mut first_bytes) {
+        Ok(_) => first_bytes == *MAGIC_BYTES,
+        Err(_) => false,
+    };
+
     let mut ledger = LEDGER.write().unwrap();
-    *ledger = ciborium::de::from_reader(&mut stable_reader).expect("Decoding stable memory failed");
+    let mut pre_upgrade_instructions_consumed = 0;
+    if !memory_manager_found {
+        // The ledger was written with dfn_core and has to be read with dfn_core in order
+        // to skip the first bytes that contain the length of the stable memory.
+        let mut stable_reader = dfn_core::stable::StableReader::new();
+        *ledger =
+            ciborium::de::from_reader(&mut stable_reader).expect("Decoding stable memory failed");
+        let mut pre_upgrade_instructions_counter_bytes = [0u8; 8];
+        pre_upgrade_instructions_consumed =
+            match stable_reader.read_exact(&mut pre_upgrade_instructions_counter_bytes) {
+                Ok(_) => u64::from_le_bytes(pre_upgrade_instructions_counter_bytes),
+                Err(_) => {
+                    // If upgrading from a version that didn't write the instructions counter to stable memory
+                    0u64
+                }
+            };
+    } else {
+        *ledger = UPGRADES_MEMORY.with_borrow(|bs| {
+            let reader = Reader::new(bs, 0);
+            let mut buffered_reader = BufferedReader::new(BUFFER_SIZE, reader);
+            let ledger_state = ciborium::de::from_reader(&mut buffered_reader).expect(
+                "Failed to read the Ledger state from memory manager managed stable structures",
```

Comment:
```suggestion
                "Failed to read the Ledger state from memory manager managed stable memory",
```
stable structures are not used yet